### PR TITLE
ci: optimize CI behavior

### LINKED ISSUE #266: ci: granular workflow
After discussing with @cedoor, we identified that the current CI workflow could be improved by taking a more granular approach. Specifically, since this repo is a monorepo, the CI triggers whenever a PR is opened, and it builds and tests every package. We could instead trigger only the workflows related to the changed packages and build and test just those packages. By reducing the overhead required to run the overall build and test steps, we could improve the CI lead time by several seconds or even minutes, which would provide much faster feedback to the contributor.

### PR DESCRIPTION
**Pull request by baumstern**
This PR introduces several optimizations to our CI pipeline.

#### Key Changes

- remove redundant dependencies
- skip unnecessary dependency installations
- disable fail-fast to continue tests after a failure; even if one fails, ensuring comprehensive test results
- cancel obsolete CI runs when a new commit is pushed to a PR

Closes #266

### CODE REVIEW

**Comment by baumstern** [comment]
The test is failing with following error 🤔

```

  115 passing (10m)
  1 failing

  1) QuinaryIMT
       # update
         Should not update a leaf that hasn't been inserted yet:
     AssertionError: Expected transaction to be reverted with custom error 'LeafIndexOutOfRange', but it reverted with custom error 'LeafDoesNotExist'
      at async Context.<anonymous> (test/QuinaryIMT.ts:192:13)
```

**Comment by cedoor** [comment]
@baumstern I think this has something to do with parallel tests and Turbo.

@sripwoud Could that be the problem here?

**Comment by baumstern** [comment]
quick heads-up:

I've added commit 3b2df4f to ensure the website build isn't broken when a PR is opened. Currently, `docs.yml` workflows are triggered only when code is pushed to the main branch

**Review by sripwoud — APPROVED** [review]
ACK

capture: #274 I think it is still worth it trying to set up the turbo cache to speed up the ci too.

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -4,19 +4,20 @@ on:
...
             - name: Checkout
-              uses: actions/checkout@v3
+              uses: actions/checkout@v4
 
             - name: Setup Node
-              uses: actions/setup-node@v3
+              uses: actions/setup-node@v4
               with:
-                  node-version: "18"
+                  node-version: "20"
```
[the remark this answers is not in the document]
Comment (cedoor) [reply .github/workflows/docs.yml:20]: Yes, I think so!

**Pull request merged**